src/minidriver/minidriver.c

```diff
@@ -1490,16 +1490,16 @@ md_set_cmapfile(PCARD_DATA pCardData, struct md_file *file)
 		 */
 		cont->size_key_exchange = cont->size_sign = 0;
 		if (key_obj->type == SC_PKCS15_TYPE_PRKEY_RSA) {
-			if (prkey_info->usage & USAGE_ANY_DECIPHER)
+			if ((prkey_info->usage & USAGE_ANY_DECIPHER) == USAGE_ANY_DECIPHER)
 				cont->size_key_exchange = prkey_info->modulus_length;
-			else if (prkey_info->usage & USAGE_ANY_SIGN)
+			else if ((prkey_info->usage & USAGE_ANY_SIGN) == USAGE_ANY_SIGN)
 				cont->size_sign = prkey_info->modulus_length;
 			else
 				cont->size_key_exchange = prkey_info->modulus_length;
 		} else if (key_obj->type == SC_PKCS15_TYPE_PRKEY_EC) {
-			if (prkey_info->usage & USAGE_ANY_AGREEMENT)
+			if ((prkey_info->usage & USAGE_ANY_AGREEMENT) == USAGE_ANY_AGREEMENT)
 				cont->size_key_exchange = prkey_info->field_length;
-			else if (prkey_info->usage & USAGE_ANY_SIGN)
+			else if ((prkey_info->usage & USAGE_ANY_SIGN) == USAGE_ANY_SIGN)
```

Comment:
Your change forces both flags to be set: `SC_PKCS15_PRKEY_USAGE_SIGN ` **and** `SC_PKCS15_PRKEY_USAGE_NONREPUDIATION`. The current logic only enforced at least one flag to be set. I think this is not intended and your changes about checking the key usage should be removed also in the other cases.